onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/providers/coreml/builders/helper.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/optimizer/initializer.h"
+
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class PadOpBuilder : public BaseOpBuilder {
+  // Add operator related
+#ifdef __APPLE__
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override;
+#endif
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+
+  int GetMinSupportedOpSet(const Node& /* node */) const override {
+    // Note: before Pad-11, inputs `pads` and `constant_value` were attributes
+    return 11;
+  }
+};
+
+// Helper function
+// Use axes initializer data if `axes` input provided or create default axes vector.
+static InlinedVector<int64_t> GetPaddingAxesData(const InitializedTensorSet& initializers,
+                                                 const Node& node, size_t input_rank) {
+  InlinedVector<int64_t> axes_tensor_data;
+  const auto& input_defs = node.InputDefs();
+
+  if (input_defs.size() > 3) {
+    // optional input axes is provided, use axes initializer data
+    const ONNX_NAMESPACE::TensorProto& axes_tensor = *initializers.at(input_defs[3]->Name());
```

Comment:
do we need to check that axes_tensor is actually an initializer first? perhaps in IsOpSupportedImpl(), and then we can assume that it is after.